app/models/miq_provision_virt_workflow.rb

```diff
@@ -336,23 +337,30 @@ def allowed_templates(options = {})
       end
     end
 
-    allowed_templates_list = source_vm_rbac_filter(vms, condition).to_a
+    # Only select the colums we need
+    vms = vms.select(:id, :name, :guid, :uid_ems, :ems_id, :cloud_tenant_id)
+
+    allowed_templates_list = source_vm_rbac_filter(vms, condition, VM_OR_TEMPLATE_EXTRA_COLS).to_a
     @allowed_templates_filter = filter_id
     @allowed_templates_tag_filters = @values[:vm_tags]
     rails_logger('allowed_templates', 1)
     if allowed_templates_list.blank?
       _log.warn("Allowed Templates is returning an empty list")
     else
       _log.warn("Allowed Templates is returning <#{allowed_templates_list.length}> template(s)")
-      allowed_templates_list.each do |vm|
-        _log.debug("Allowed Template <#{vm.id}:#{vm.name}>  GUID: <#{vm.guid}>  UID_EMS: <#{vm.uid_ems}>")
+      if _log.level == Logger::DEBUG # skip .each if not in DEBUG
+        allowed_templates_list.each do |vm|
+          _log.debug("Allowed Template <#{vm.id}:#{vm.name}>  GUID: <#{vm.guid}>  UID_EMS: <#{vm.uid_ems}>")
+        end
       end
     end
 
-    MiqPreloader.preload(allowed_templates_list, [:snapshots, :operating_system, :ext_management_system, {:hardware => :disks}])
+    MiqPreloader.preload(allowed_templates_list, %i(operating_system))
+    @_ems_allowed_templates_cache = {}
```

Comment:
is there a way to repopulate this?

Although, there will probably be only 1 ems.

Reply:
> is there a way to repopulate this?

Not sure what you are asking.  Can you explain further?

> Although, there will probably be only 1 ems.

Yes, that is the main idea.  With the scale we were working with (120k VM templates), using this cache allows us to re-use the loop iteration that is done for instantiating the `MiqHashStruct` objects instead of doing a separate one for the preload.  This technically means we are doing a "N+1", but N in this case is literally 1 anyway (and should almost always be), so we should be fine with that change.

Reply:
I was hoping we would know the ems and pre-populate this hash. But think the answer to my question was that we can not.

Reply:
> I was hoping we would know the ems and pre-populate this hash.

Ah, I see (pre-populate not "repopulate").

The no, there is no way to do this **AND** avoid doing another iteration through the list to find what those IDs were in the first place.  If I wasn't concerned about another loop iteration (again, 120k records here), then I would have just used preload and called it a day since we could just add a scope to it like this:

```ruby
MiqPreloader.preload(allowed_templates_list, %i(operating_system))
# doing two of these totally works
MiqPreloader.preload(allowed_templates_list, %i(ext_management_system), ExtManagementSystem.select(:name))
```

If, however, you think this is a performance gains our outweighed by the obscurity of this code, I can understand not wanting to move forward with this.  Thoughts?